File: docs/site/samples/python/examples/guides_guard_route_multi.py
"""Routing the guard text for a multi-message composition: prepend the advisory
as a system message, then the resolved body messages. Standalone — run directly
or under pytest."""

from pathlib import Path

from prompting_press import Prompt, GuardConfig, Composition
from pydantic import BaseModel, Field

# The caller reads the definition; the library does no file I/O itself.
# Resolve the file next to this program (a real app uses its own path).
_HERE = Path(__file__).parent
ask = Prompt.from_yaml((_HERE / "ask.yaml").read_text())


class Ask(BaseModel):
    topic: str = Field(min_length=1)


def main() -> None:
    # A single-render result supplies the guard advisory to prepend.
    result = ask.render(Ask, data={"topic": "rivers"}, guard=GuardConfig(enabled=True))

    # The body messages come from a composition.
    comp = Composition()
    comp.append(ask, Ask(topic="rivers"))

    guard_msg = {"role": "system", "content": result.guard}
    body_messages = [{"role": m.role, "content": m.text} for m in comp.resolve()]
    request_messages = [guard_msg] + body_messages

    assert request_messages[0]["role"] == "system"
    assert "<untrusted>" in request_messages[0]["content"]
    assert request_messages[1]["role"] == "user"
    # Composition never applies the guard, so its body is the plain render.
    assert request_messages[1]["content"] == "Tell me about rivers."


if __name__ == "__main__":
    main()

File: docs/site/samples/typescript/examples/guides_guard_route_multi.ts
// Routing the guard text for a multi-message composition: prepend the advisory
// as a system message, then the resolved body messages. Standalone — run under
// `node --test`.

import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { test } from "node:test";
import { Composition, Prompt } from "prompting-press";
import { z } from "zod";

// The caller reads the definition; the library does no file I/O itself.
// Resolve the file next to this program (a real app uses its own path).
const defFile = (name: string) => fileURLToPath(new URL(name, import.meta.url));

const ask = Prompt.fromYaml(readFileSync(defFile("ask.yaml"), "utf8"));

const Ask = z.object({ topic: z.string().min(1) });

test("prepend the guard advisory to the resolved composition body", () => {
	// A single-render result supplies the guard advisory to prepend.
	const result = ask.render(
		Ask,
		{ topic: "rivers" },
		{ guard: { enabled: true } },
	);

	// The body messages come from a composition.
	const comp = new Composition();
	comp.append({ prompt: ask, schema: Ask, data: { topic: "rivers" } });

	const guardMsg = { role: "system", content: result.guard };
	const bodyMessages = comp
		.resolve()
		.map((m) => ({ role: m.role, content: m.text }));
	const requestMessages = [guardMsg, ...bodyMessages];

	assert.equal(requestMessages[0].role, "system");
	assert.ok(requestMessages[0].content?.includes("<untrusted>"));
	assert.equal(requestMessages[1].role, "user");
	// Composition never applies the guard, so its body is the plain render.
	assert.equal(requestMessages[1].content, "Tell me about rivers.");
});
